README.md
# Arduino TOOLBOX

Arduino Communication through Serial

## Getting Started


### Prerequisites

1. Scilab >= 2023.0.0

### Installing

The module could be easily install from the Scilab atoms portal by using command:

--> atomsInstall('arduino'); // Coming soon

or the offline version by replacing the input argument to the zip file name.

If you prefer to build then module, download the source and unzip it, just type in scilab:

```
--> exec("arduino_Path/builder.sce")
```

If the module is place in the "contrib" folder, it could be loaded from the "Toolboxes" menu. Otherwise, just type in scilab:

``` 
--> exec("arduino_path/loader.sce"); 
```

arduino_Path is where this file is.

### EXAMPLES and DEMOS
 To see some examples type in scilab:

--> help 

## Features
* Analog input
* Analog output (PWM)
* Digital input
* Digital output
* Encoder input
* Interrupt
* DC motor
* Servo motor

## Revisions

### REVISION NOTES 1.9
BUGS FIXED
* Fixed issue of get_figure_handle removed in Scilab 2023.0.0

### REVISION NOTES 1.8
BUGS FIXED
* Fixed issue of extra figure popup when demo figure is on

NEW FEATURES
* Add command lines control support for common interface.
* Translate help pages in english
* Translate demos

### REVISION NOTES 1.7.1
BUGS FIXED
* Fixed 2206 Problem with Servo_Write Block
* Restore MPU6050 block

### REVISION NOTES 1.7
BUGS FIXED
* Fixed Scilab 6 compatibility issues
* Change the serial to Scilab from char to double to solved ascii issues.
* Fixed port >9 not supported issue
  
NEW FEATURES
* Multiple boards support (Set to 2 currently as more board yields slow response)
* Linux/MacOS(?) Support (Thanks to Nikolay Strelkov link in http://forge.scilab.org/index.php/p/arduino/issues/1556/) 
* Modified the dll for serial communication to Scilab 6.1.x gateway format, allowing other contributors to enhance it easier.

## To-Do
   * Change the firmware to get multiple channels and sent all back to host for performance improvement
   * Translate help pages in english
   * Translate blocks interfaces and error messages
   * Translate demos
 

## Version

This is the Revision 1.9 of arduino toolbox for Scilab >= 2023.0.0

## Authors

This is the Revision 1.9 of arduino toolbox for Scilab >= 2023.0.0
Modified By Tan Chi Luh 

Original Works By Bruno JOFRET

## License

This program is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation; either version 2 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with this program; if not, write to the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

## Acknowledgments

1. Bruno Jofret 
2. Yann Debray 
3. Samuel Gougeon

## Extra Demo Videos
1. Using Scilab arduino module for soft real-time closed-loop control
https://www.linkedin.com/posts/tanchinluh_scilab-activity-6590880449571192832-FwU9
2. Scilab arduino under linux
https://www.linkedin.com/posts/tanchinluh_scilab-activity-6590139219363094528-OAvt
3. Working with 2 Ardiono Uno boards
https://www.linkedin.com/posts/tanchinluh_scilab-arduino-activity-6589898816311390208-Oo7Z
https://www.linkedin.com/posts/tanchinluh_scilab-arduino-activity-6589484169330561024-A3io



changelog.txt
==========================
Arduino toolbox for Scilab
==========================

v. 1.9 (2023-10-11) / stephane.mottelet@utc.fr
BUGS FIXED
* Fixed issue of get_figure_handle removed in Scilab 2023.0.0

v. 1.8 (2020-02-16) / chinluh.tan@bytecode-asia.com
BUGS FIXED
* Fixed issue of extra figure popup when demo figure is on

NEW FEATURES
* Add command lines control support for common interface.
* Translate help pages in english
* Translate demos

v. 1.7.1 (2020-01-05) / chinluh.tan@bytecode-asia.com
BUGS FIXED
* Fixed 2206 Problem with Servo_Write Block
* Restore MPU6050 block

v. 1.7 (2019-10-17) / chinluh.tan@bytecode-asia.com
BUG FIXED
  * Fixed Scilab 6 compatibility issues
  * Change the serial to Scilab from char to double to solved ascii issues.
  * Fixed port >9 not supported issue
  
  NEW FEATURES
  * Multiple boards support (Set to 2 currently as more board yields slow response)
  * Linux/MacOS(?) Support (Thanks to Nikolay Strelkov link in http://forge.scilab.org/index.php/p/arduino/issues/1556/) 

v. 1.6.2 (2019-06-04) / sgougeon@free.fr
---------------------
 BUG FIXED
 * ARDUINO_pre_simulate() was still too stiff to detect the version answer from the board


v. 1.6 (2019-05-30) / sgougeon@free.fr
-------------------
 BUGS FIXED
 * toolbox_arduino_v4.ino file shipped in the module was bad. Replaced with the right file.
 * ARDUINO_pre_simulate() was expecting arduino_v3.ino instead of arduino_v4.ino
   and yielded an error "Il faut charger avec le logiciel arduino le sketch
   toolbox_arduino_v3.ino dans la carte Arduino" when actually dealing with the v.4
 * The DESCRIPTION file that was removed from v.1.4 is actually needed to instal
   the module offline. Restored.
 * Demos: Examples for Scilab 6.0 are displayed with wrong ports positions with Scilab 5.5.
   The versions for Arduino 1.1 are restored in the binary for the Scilab 5.5.


v. 1.5 (2018-09-08) / sgougeon@free.fr
-------------------
* toolbox_arduino_v4.ino is now shipped within the toolbox, including the
  MPU6050 bloc and filtering. Thanks to David VIOLEAU.
* etc/arduino-style.xml is partly provided (but it does not work).

BUGS FIXED
* Icons labels are restored and improved.
* All demos are restored.
* In the example9.zcos demo, the PI block refered to an inexisting PIcontrol
  function.
* The arduino.start changed the user working directory. Fixed.
* init_arduino() and ARDUINO_post_simulate() had still some old cells


v. 1.4 (2018-09-07) / sgougeon@free.fr
-------------------
This is mainly a repackaging version.

* The builder can now build for both Scilab 5 and Scilab 6
* HELP PAGES
  - images are now gathered in a dedicated directory
  - .tst source files are now gathered in a dedicated directory
  - _README.tst file added: only XML files must be updated/upgraded/translated
  - See also sections: the short descriptions of items are now displayed
  - Contents sections simplified
* DESCRIPTION FILE removed (no longer needed)
* changelog.txt added (this file)
* bin directory created
* BUGS FIXED:
  - the scicos_scicoslib was missing when loading the Arduino module
  - getArduinoVersion() was not up-to-date. It is no longer necessary to update
    it: It now reads the VERSION file.
  - In macros, some messages were in french.
  
TODO:
  * Extend the module to Linux users (through the serial toolbox based on TCL)
  * Translate help pages in english
  * Translate error blocks interfaces and messages
  * Translate demos


sci_gateway/c/main_linux.c
/**************************************************
file: main.c
purpose: serial librairie for Scilab

Alain Caignot
**************************************************/

#include <errno.h>
#include <fcntl.h>
#include <string.h>
#include <sys/ioctl.h>
#include <termios.h>
#include <unistd.h>

#include "api_scilab.h"
#include "Scierror.h"
#include "sciprint.h"

static scilabStatus set_interface_attribs(int fd, int baudrate, int parity)
{
	speed_t speed;
	struct termios tty;

	if (tcgetattr(fd, &tty) < 0)
	{
		return STATUS_ERROR;
	}

	switch(baudrate)
	{
	case 115200:
		speed = B115200;
		break;
	default:
		return STATUS_ERROR;
	}

	if (cfsetospeed(&tty, speed) < 0 || cfsetispeed(&tty, speed) < 0)
	{
		return STATUS_ERROR;
	}

	tty.c_cflag = (tty.c_cflag & ~CSIZE) | CS8; // 8-bit chars
	// disable IGNBRK for mismatched speed tests; otherwise receive break
	// as \000 chars
	tty.c_iflag &= ~IGNBRK; // disable break processing
	tty.c_lflag = 0; // no signaling chars, no echo,
	// no canonical processing
	tty.c_oflag = 0; // no remapping, no delays
	tty.c_cc[VMIN] = 0; // read doesn't block
	tty.c_cc[VTIME] = 5; // 0.5 seconds read timeout

	tty.c_iflag &= ~(IXON | IXOFF | IXANY); // shut off xon/xoff ctrl

	tty.c_cflag |= (CLOCAL | CREAD); // ignore modem controls,
	// enable reading
	tty.c_cflag &= ~(PARENB | PARODD);// shut off parity
	tty.c_cflag |= parity;
	tty.c_cflag &= ~CSTOPB;
	tty.c_cflag &= ~CRTSCTS;

	if (tcsetattr(fd, TCSANOW, &tty) < 0)
	{
		return STATUS_ERROR;
	}

	return STATUS_OK;
}

static scilabStatus set_blocking(int fd, int should_block)
{
	struct termios tty;

	if (tcgetattr(fd, &tty) < 0)
	{
		return STATUS_ERROR;
	}

	tty.c_cc[VMIN] = should_block ? 1 : 0;
	tty.c_cc[VTIME] = 5; // 0.5 seconds read timeout

	if (tcsetattr(fd, TCSANOW, &tty) < 0)
	{
		return STATUS_ERROR;
	}

	return STATUS_OK;
}

int open_serial(scilabEnv env, int nin, scilabVar* in, int nopt, scilabOpt* opt, int nout, scilabVar* out)
{
	wchar_t *in1;
	double in2;
	int fd;
	int baudrate;

	(void)nopt;
	(void)opt;

	// Check the number of input arguments
	if (nin != 3)
	{
		Scierror(77, "Wrong number of input argument(s): %d expected.\n", 3);
		return STATUS_ERROR;
	}

	// Check the number of output arguments
	if (nout != 1)
	{
		Scierror(77, "Wrong number of output arguments: %d expected.\n", 1);
		return STATUS_ERROR;
	}

	//in[0] : port handle number - ignored

	//in[1] : String
	if (scilab_isString(env, in[1]) == 0 || scilab_isScalar(env, in[1]) == 0)
	{
		Scierror(77, "Wrong type for input argument %d: A string expected.\n", 2);
		return STATUS_ERROR;
	}
	scilab_getString(env, in[1], &in1);

	//in[2] : Double
	if (scilab_isDouble(env, in[2]) == 0 || scilab_isScalar(env, in[2]) == 0)
	{
		Scierror(77, "Wrong type for input argument %d: A double expected.\n", 3);
		return STATUS_ERROR;
	}
	scilab_getDouble(env, in[2], &in2);
	baudrate = (int)(in2 + 0.5);

	char portname[100];
	wcstombs(portname, in1, sizeof(portname));

	fd = open(portname, O_RDWR | O_NOCTTY | O_SYNC);
	//fd = open(portname, O_RDWR | O_NOCTTY); //srikant
	if (fd < 0)
	{
		Scierror(999, "Fail to open serial port %s: %s\n", portname, strerror(errno));
		return STATUS_ERROR;
	}

	//sciprint("Opened port %d: %s at %d\n", fd, portname, baudrate);

	if (set_interface_attribs(fd, baudrate, 0) != STATUS_OK ||
		set_blocking(fd, 0) != STATUS_OK)
	{
		Scierror(999, "Fail to configure port %s: %s\n", portname, strerror(errno));
		return STATUS_ERROR;
	}

	// Clear BREAK
	ioctl(fd, TIOCCBRK, NULL);

	// Reset DTR and RTS
	int flags = TIOCM_DTR | TIOCM_RTS;
	ioctl(fd, TIOCMBIC, &flags);

	out[0] = scilab_createPointer(env, (void*)(intptr_t)fd);

	sleep(1);
	return STATUS_OK;
}

int close_serial(scilabEnv env, int nin, scilabVar* in, int nopt, scilabOpt* opt, int nout, scilabVar* out)
{
	void *in0;
	int fd;

	(void)nopt;
	(void)opt;
	(void)out;

	// Check number of input arguments
	if (nin != 1)
	{
		Scierror(77, "Wrong number of input argument(s): %d expected.\n", 1);
		return STATUS_ERROR;
	}

	// Check the number of output arguments
	if (nout != 0)
	{
		Scierror(77, "Wrong number of output arguments: %d expected.\n", 0);
		return STATUS_ERROR;
	}

	// in[0] : pointer
	if (scilab_isPointer(env, in[0]) == 0)
	{
		Scierror(999, "Wrong type for input argument %d: A pointer expected.\n", 1);
		return STATUS_ERROR;
	}
	scilab_getPointer(env, in[0], &in0);
	fd = (intptr_t)in0;

	//sciprint("Closing port %d\n", fd);

	if (close(fd) != 0)
	{
		sciprint("Failed to close the port: %s\n", strerror(errno));
	}

	return STATUS_OK;
}

int write_serial(scilabEnv env, int nin, scilabVar* in, int nopt, scilabOpt* opt, int nout, scilabVar* out)
{
	void *in0;
	wchar_t* in1;
	double in2;
	int fd;
	int size;

	(void)nopt;
	(void)opt;
	(void)out;

	// Check number of input arguments
	if (nin != 3)
	{
		Scierror(77, "Wrong number of input argument(s): %d expected.\n", 3);
		return STATUS_ERROR;
	}

	// Check the number of output arguments
	if (nout != 0)
	{
		Scierror(77, "Wrong number of output arguments: %d expected.\n", 0);
		return STATUS_ERROR;
	}

	// in[0] : pointer
	if (scilab_isPointer(env, in[0]) == 0)
	{
		Scierror(999, "%s: Wrong type for input argument %d: A pointer expected.\n", __func__, 1);
		return STATUS_ERROR;
	}
	scilab_getPointer(env, in[0], &in0);
	fd = (intptr_t)in0;

	// in[1] : string
	if (scilab_isString(env, in[1]) == 0 || scilab_isScalar(env, in[1]) == 0)
	{
		Scierror(999, "Wrong type for input argument %d: A String expected.\n", 2);
		return STATUS_ERROR;
	}
	scilab_getString(env, in[1], &in1);

	// in[2] : Double
	if (scilab_isDouble(env, in[2]) == 0 || scilab_isScalar(env, in[2]) == 0)
	{
		Scierror(999, "Wrong type for input argument %d: A double expected.\n", 3);
		return STATUS_ERROR;
	}
	scilab_getDouble(env, in[2], &in2);
	size = (int)(in2 + 0.5);

	//sciprint("Writing %d bytes to port %d\n", size, fd);

	// Convert input string from wchar to char
	char ch[100];
	wcstombs(ch, in1, sizeof(ch));

	int ret = write(fd, ch, size);

	if (ret < 0)
	{
		Scierror(999, "Write to port failed: %s\n", strerror(errno));
		return STATUS_ERROR;
	}

	if (ret != size)
	{
		Scierror(999,"Unexpected number of bytes written: expected %d but written %d", size, ret);
		return STATUS_ERROR;
	}

	return STATUS_OK;
}

int status_serial(scilabEnv env, int nin, scilabVar* in, int nopt, scilabOpt* opt, int nout, scilabVar* out)
{
	void *in0;
	int fd;

	(void)nopt;
	(void)opt;

	// Check number of input arguments
	if (nin != 1)
	{
		Scierror(77, "Wrong number of input argument(s): %d expected.\n", 1);
		return STATUS_ERROR;
	}

	// Check the number of output arguments
	if (nout != 3)
	{
		Scierror(77, "Wrong number of output arguments: %d expected.\n", 3);
		return STATUS_ERROR;
	}

	// in[0] : pointer
	if (scilab_isPointer(env, in[0]) == 0)
	{
		Scierror(999, "%s: Wrong type for input argument %d: A pointer expected.\n", __func__, 1);
		return STATUS_ERROR;
	}
	scilab_getPointer(env, in[0], &in0);
	fd = (intptr_t)in0;

	int nbread = 0;
	int nbwrite = 0;

	ioctl(fd, TIOCINQ, &nbread);
	ioctl(fd, TIOCOUTQ, &nbwrite);

	//sciprint("Status port %d: in=%d out=%d\n", fd, nbread, nbwrite);

	// Create outputs
	out[0] = scilab_createDouble(env, (double)0);
	out[1] = scilab_createDouble(env, (double)nbread);
	out[2] = scilab_createDouble(env, (double)nbwrite);

	return STATUS_OK;
}

int read_serial(scilabEnv env, int nin, scilabVar* in, int nopt, scilabOpt* opt, int nout, scilabVar* out)
{
	void *in0;
	double in1;
	int fd;
	int size;
	unsigned char buf[200];

	(void)nopt;
	(void)opt;

	// Check number of input arguments
	if (nin != 2)
	{
		Scierror(77,"Wrong number of input argument(s): %d expected.\n", 2);
		return STATUS_ERROR;
	}

	// Check the number of output arguments
	if (nout != 1)
	{
		Scierror(77, "Wrong number of output arguments: %d expected.\n", 1);
		return STATUS_ERROR;
	}

	// in[0] : pointer
	if (scilab_isPointer(env, in[0]) == 0)
	{
		Scierror(999, "%s: Wrong type for input argument #%d: A pointer expected.\n", __func__, 1);
		return STATUS_ERROR;
	}
	scilab_getPointer(env, in[0], &in0);
	fd = (intptr_t)in0;

	// in[1] : double
	if (scilab_isDouble(env, in[1]) == 0 || scilab_isScalar(env, in[1]) == 0)
	{
		Scierror(999,"Wrong type for input argument %d: A integer expected.\n", 2);
		return STATUS_ERROR;
	}
	scilab_getDouble(env, in[1], &in1);
	size = (int)(in1 + 0.5);

	//sciprint("Read %d bytes from port %d\n", size, fd);

	int ret = read(fd, buf, size);

	if (ret < 0)
	{
		Scierror(999, "Read from port failed: %s\n", strerror(errno));
		return STATUS_ERROR;
	}

	if (ret != size)
	{
		Scierror(999,"Unexpected number of bytes read: expected %d but read %d", size, ret);
		return STATUS_ERROR;
	}

	// Creating output as double
	double* out1 = NULL;

	out[0] = scilab_createDoubleMatrix2d(env, 1, size, 0);
	scilab_getDoubleArray(env, out[0], &out1);

	for (int i = 0; i < size; ++i)
	{
		out1[i] = (double)*(buf+i);
	}

	return STATUS_OK;
}


sci_gateway/c/main_win.c
/**************************************************
file: main.c
purpose: serial librairie for Scilab

Alain Caignot
**************************************************/

#include <windows.h>
#include <string.h>
#include <stdio.h>

#include "api_scilab.h"
#include "Scierror.h"
#include "sciprint.h"

static LPVOID GetLastErrorString(void)
{
	LPVOID *lpLpMsgBuf = NULL;

	FormatMessage(
		FORMAT_MESSAGE_ALLOCATE_BUFFER |
		FORMAT_MESSAGE_FROM_SYSTEM |
		FORMAT_MESSAGE_IGNORE_INSERTS,
		NULL,
		GetLastError(),
		MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT),
		(LPTSTR) &lpLpMsgBuf,
		0,
		NULL);

	return lpLpMsgBuf;
}

static scilabStatus GetFullPortName(char *name, size_t maxlen)
{
	char *number = name;
	char *endp;

	if (strncmp(name, "COM", 3) == 0)
		number = name + 3;

	long n = strtol(number, &endp, 10);
	if (n < 1 || n > 255 || name == endp || *endp != '\0')
		return STATUS_ERROR;

	int len = snprintf(name, maxlen, "\\\\.\\COM%d", n);
	if (len >= maxlen)
		return STATUS_ERROR;

	return STATUS_OK;
}

int open_serial(scilabEnv env, int nin, scilabVar* in, int nopt, scilabOpt* opt, int nout, scilabVar* out)
{
	DCB dcbSerialParams;
	COMMTIMEOUTS timeouts = { 0 };
	wchar_t *in1;
	double in2;
	void *hPort;
	int baudrate;

	(void)nopt;
	(void)opt;

	// Check the number of input arguments
	if (nin != 3)
	{
		Scierror(77, "Wrong number of input argument(s): %d expected.\n", 3);
		return STATUS_ERROR;
	}

	// Check the number of output arguments
	if (nout != 1)
	{
		Scierror(77, "Wrong number of output arguments: %d expected.\n", 1);
		return STATUS_ERROR;
	}

	//in[1] : String
	if (scilab_isString(env, in[1]) == 0 || scilab_isScalar(env, in[1]) == 0)
	{
		Scierror(77, "Wrong type for input argument %d: A string expected.\n", 2);
		return STATUS_ERROR;
	}
	scilab_getString(env, in[1], &in1);

	//in[2] : Double
	if (scilab_isDouble(env, in[2]) == 0 || scilab_isScalar(env, in[2]) == 0)
	{
		Scierror(77, "Wrong type for input argument %f: A double expected.\n", 3);
		return STATUS_ERROR;
	}
	scilab_getDouble(env, in[2], &in2);
	baudrate = (int)(in2 + 0.5);

	char portname[100];
	wcstombs(portname, in1, sizeof(portname));

	if (GetFullPortName(portname, sizeof(portname)) != STATUS_OK)
	{
		Scierror(77, "Invalid port name '%s'. Expected a number in range [1,255] with optional COM prefix", portname);
		return STATUS_ERROR;
	}

	// Create Serial port handle
	hPort = CreateFile(portname,
		GENERIC_READ | GENERIC_WRITE,
		0,//FILE_SHARE_READ | FILE_SHARE_WRITE //to test : recuperation COM port if simulation crashes
		0,
		OPEN_EXISTING,
		FILE_ATTRIBUTE_NORMAL,
		0);

	//sciprint("Open %s at %d bps %p\n", portname, baudrate, hPort);

	// Errors checking
	if (hPort == INVALID_HANDLE_VALUE)
	{
		char *lpMsg = GetLastErrorString();
		Scierror(999, "Failed to open %s: %s\n", portname, lpMsg);
		LocalFree(lpMsg);
		return STATUS_ERROR;
	}

	dcbSerialParams.DCBlength = sizeof(dcbSerialParams);
	if (!GetCommState(hPort, &dcbSerialParams))
	{
		char *lpMsg = GetLastErrorString();
		Scierror(999, "Error getting state: %s\n", lpMsg);
		LocalFree(lpMsg);
		return STATUS_ERROR;
	}
	dcbSerialParams.BaudRate = baudrate;
	dcbSerialParams.ByteSize = 8;
	dcbSerialParams.StopBits = ONESTOPBIT;
	dcbSerialParams.Parity = NOPARITY;
	dcbSerialParams.fOutxCtsFlow = FALSE;
	dcbSerialParams.fOutxDsrFlow = FALSE;
	dcbSerialParams.fDtrControl = DTR_CONTROL_DISABLE;
	dcbSerialParams.fRtsControl = RTS_CONTROL_DISABLE;
	dcbSerialParams.fOutX = FALSE;
	dcbSerialParams.fInX = FALSE;
	if (!SetCommState(hPort, &dcbSerialParams))
	{
		char *lpMsg = GetLastErrorString();
		Scierror(999, "Error setting state with error\n", lpMsg);
		LocalFree(lpMsg);
		return STATUS_ERROR;
	}

	//
	timeouts.ReadIntervalTimeout = 50;
	timeouts.ReadTotalTimeoutConstant = 50;
	timeouts.ReadTotalTimeoutMultiplier = 1;
	timeouts.WriteTotalTimeoutConstant = 50;
	timeouts.WriteTotalTimeoutMultiplier = 1;
	if (!SetCommTimeouts(hPort, &timeouts))
	{
		char *lpMsg = GetLastErrorString();
		Scierror(999, "Failed to configure timeouts: %s\n", lpMsg);
		LocalFree(lpMsg);
		return STATUS_ERROR;
	}

	if (!EscapeCommFunction(hPort, CLRBREAK) ||
		!EscapeCommFunction(hPort, CLRDTR) ||
		!EscapeCommFunction(hPort, CLRRTS))
	{
		// Failed to clear BREAK, DTR and RTS
		char *lpMsg = GetLastErrorString();
		Scierror(999, "Failed to reset signals with error\n", lpMsg);
		LocalFree(lpMsg);
		return STATUS_ERROR;
	}

	//sciprint("Serial port %s opened SUCCESSFULLY.\n", portname);
	out[0] = scilab_createPointer(env, hPort);

	Sleep(1000);
	return STATUS_OK;
}

int close_serial(scilabEnv env, int nin, scilabVar* in, int nopt, scilabOpt* opt, int nout, scilabVar* out)
{
	void *hPort;

	(void)nopt;
	(void)opt;
	(void)out;

	// Check number of input arguments
	if (nin != 1)
	{
		Scierror(77, "Wrong number of input argument(s): %d expected.\n", 1);
		return STATUS_ERROR;
	}

	// Check the number of output arguments
	if (nout != 0)
	{
		Scierror(77, "Wrong number of output arguments: %d expected.\n", 0);
		return STATUS_ERROR;
	}

	// in[0] : pointer
	if (scilab_isPointer(env, in[0]) == 0)
	{
		Scierror(999, "Wrong type for input argument %d: A pointer expected.\n", 1);
		return STATUS_ERROR;
	}
	scilab_getPointer(env, in[0], &hPort);

	//sciprint("Close port %p\n", hPort);

	// Check whether the port is closed successfully.
	if (!CloseHandle(hPort))
	{
		char *lpMsg = GetLastErrorString();
		Scierror(999, "Failed to close the port: %s\n", lpMsg);
		LocalFree(lpMsg);
		return STATUS_ERROR;
	}

	return STATUS_OK;
}

int write_serial(scilabEnv env, int nin, scilabVar* in, int nopt, scilabOpt* opt, int nout, scilabVar* out)
{
	DWORD dwBytesWrite;
	void *hPort;
	wchar_t* in1;
	double in2;
	DWORD size;

	(void)nopt;
	(void)opt;
	(void)out;

	// Check number of input arguments
	if (nin != 3)
	{
		Scierror(77, "Wrong number of input argument(s): %d expected.\n", 3);
		return STATUS_ERROR;
	}

	// Check the number of output arguments
	if (nout != 0)
	{
		Scierror(77, "Wrong number of output arguments: %d expected.\n", 0);
		return STATUS_ERROR;
	}

	// in[0] : pointer
	if (scilab_isPointer(env, in[0]) == 0)
	{
		Scierror(999, "Wrong type for input argument %d: A pointer expected.\n", 1);
		return STATUS_ERROR;
	}
	scilab_getPointer(env, in[0], &hPort);

	// in[1] : string
	if (scilab_isString(env, in[1]) == 0 || scilab_isScalar(env, in[1]) == 0)
	{
		Scierror(999, "Wrong type for input argument %d: A String expected.\n", 2);
		return STATUS_ERROR;
	}
	scilab_getString(env, in[1], &in1);

	// in[2] : Double
	if (scilab_isDouble(env, in[2]) == 0 || scilab_isScalar(env, in[2]) == 0)
	{
		Scierror(999, "Wrong type for input argument %d: A double expected.\n", 3);
		return STATUS_ERROR;
	}
	scilab_getDouble(env, in[2], &in2);
	size = (DWORD)(in2 + 0.5);

	//sciprint("Write %d bytes into port %p\n", size, hPort);

	// Convert input string from wchar to char
	char ch[100];
	wcstombs(ch, in1, sizeof(ch));

	if (!WriteFile(hPort, ch, size, &dwBytesWrite, NULL))
	{
		char *lpMsg = GetLastErrorString();
		Scierror(999,"Failed to write into port: %s", lpMsg);
		LocalFree(lpMsg);
		return STATUS_ERROR;
	}

	if (size != dwBytesWrite)
	{
		Scierror(999,"Unexpected number of bytes written: expected %d but written %d", size, dwBytesWrite);
		return STATUS_ERROR;
	}

	return STATUS_OK;
}

int status_serial(scilabEnv env, int nin, scilabVar* in, int nopt, scilabOpt* opt, int nout, scilabVar* out)
{
	DWORD dwErrorFlags;
	COMSTAT ComStat;
	void *hPort;
	int nbread;
	int nbwrite;

	(void)nopt;
	(void)opt;

	// Check number of input arguments
	if (nin != 1)
	{
		Scierror(77, "Wrong number of input argument(s): %d expected.\n", 1);
		return STATUS_ERROR;
	}

	// Check the number of output arguments
	if (nout != 3)
	{
		Scierror(77, "Wrong number of output arguments: %d expected.\n", 3);
		return STATUS_ERROR;
	}

	// in[0] : pointer
	if (scilab_isPointer(env, in[0]) == 0)
	{
		Scierror(999, "Wrong type for input argument %d: A pointer expected.\n", 1);
		return STATUS_ERROR;
	}
	scilab_getPointer(env, in[0], &hPort);

	if (!ClearCommError(hPort, &dwErrorFlags, &ComStat))
	{
		char *lpMsg = GetLastErrorString();
		Scierror(999,"Failed to get port's status: %s\n", lpMsg);
		LocalFree(lpMsg);
		return STATUS_ERROR;
	}

	nbread  = ComStat.cbInQue;
	nbwrite = ComStat.cbOutQue;

	//sciprint("Status port %p: in=%d out=%d\n", hPort, nbread, nbwrite);

	// Create outputs
	out[0] = scilab_createDouble(env, (double)0);
	out[1] = scilab_createDouble(env, (double)nbread);
	out[2] = scilab_createDouble(env, (double)nbwrite);

	return STATUS_OK;
}

int read_serial(scilabEnv env, int nin, scilabVar* in, int nopt, scilabOpt* opt, int nout, scilabVar* out)
{
	DWORD dwBytesRead;
	void *hPort;
	double in1;
	DWORD size;
	unsigned char buf[10];

	(void)nopt;
	(void)opt;

	// Check number of input arguments
	if (nin != 2)
	{
		Scierror(77,"Wrong number of input argument(s): %d expected.\n", 2);
		return STATUS_ERROR;
	}

	// Check the number of output arguments
	if (nout != 1)
	{
		Scierror(77, "Wrong number of output arguments: %d expected.\n", 1);
		return STATUS_ERROR;
	}

	// in[0] : pointer
	if (scilab_isPointer(env, in[0]) == 0)
	{
		Scierror(999, "Wrong type for input argument #%d: A pointer expected.\n", 1);
		return STATUS_ERROR;
	}
	scilab_getPointer(env, in[0], &hPort);

	// in[1] : double
	if (scilab_isDouble(env, in[1]) == 0 || scilab_isScalar(env, in[1]) == 0)
	{
		Scierror(999,"Wrong type for input argument %d: A double expected.\n", 2);
		return STATUS_ERROR;
	}
	scilab_getDouble(env, in[1], &in1);
	size = (DWORD)(in1 + 0.5);

	if (!ReadFile(hPort, buf, size, &dwBytesRead, NULL))
	{
		char *lpMsg = GetLastErrorString();
		Scierror(999,"Failed to read from port: %s\n", lpMsg);
		LocalFree(lpMsg);
		return STATUS_ERROR;
	}

	if (size != dwBytesRead)
	{
		Scierror(999,"Unexpected number of bytes read: expected %d but read %d", size, dwBytesRead);
		return STATUS_ERROR;
	}

	//sciprint("Read %d bytes from port %p\n", size, hPort);

	// Creating output as double
	double* out1;

	out[0] = scilab_createDoubleMatrix2d(env, 1, size, 0);
	scilab_getDoubleArray(env, out[0], &out1);

	for (unsigned i = 0; i < dwBytesRead; ++i)
	{
		out1[i] = (double)*(buf+i);
	}

	return STATUS_OK;
}
